### PR TITLE
Addition of MADGRAD Optimizer Algorithm

### PR DESCRIPTION
**Pull request by DarshanDeshpande**
# Description
The MADGRAD algorithm is a reliably high-performing non-convex optimization algorithm that matches or outperforms the SGD and Adam algorithms on a variety of image-to-image and language tasks. This PR includes a tensorflow-based implementation of the same.

Brief Description of the PR:
The PR adds the MadGrad class as a sub-class of OptimizerV2 and tests related to it.

Fixes # (issue)
[#2452 ](https://github.com/tensorflow/addons/issues/2452)

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [X] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [X] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [X] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [X] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?
Many tests for sparse and dense vector optimization have been performed. ```madgrad_test``` also includes a tf.keras-based example to help with design clarification. There are no tests for ```tf.keras.optimizers.schedules``` support, but this MadGrad implementation supports schedulers out of the box and the tests can be included if necessary.

### CODE REVIEW

File: tensorflow_addons/optimizers/madgrad.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+# limitations under the License.
+# ==============================================================================
+"""Madgrad optimizer implementation."""
+# pylint: disable=g-classes-have-attributes
+
+from tensorflow.python.framework import ops
+from tensorflow.python.keras import backend_config
+from tensorflow.python.keras.optimizer_v2 import optimizer_v2
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import math_ops
```
Comment (fsx950223) [inline tensorflow_addons/optimizers/madgrad.py:22]: Use public apis instead of internal apis.

Comment (DarshanDeshpande) [reply tensorflow_addons/optimizers/madgrad.py]: Done. Switched to public apis

File: tensorflow_addons/optimizers/madgrad.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        name="Madgrad",
+        **kwargs,
+    ):
+        learning_rate = kwargs.get("lr", learning_rate)
+        super(MadGrad, self).__init__(name, **kwargs)
+        self._set_hyper("learning_rate", kwargs.get("lr", learning_rate))
+        self._set_hyper("decay", self._initial_decay)
+        self._set_hyper("momentum", momentum)
+        self._set_hyper("weight_decay", weight_decay)
+        self.epsilon = epsilon or backend_config.epsilon()
```
Comment (fsx950223) [inline tensorflow_addons/optimizers/madgrad.py:69]: tf.keras.backend.epsilon()

Comment (DarshanDeshpande) [reply tensorflow_addons/optimizers/madgrad.py]: Done. Please check.

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
